drivers/platform/stm32/stm32_spi.c

```diff
@@ -227,6 +227,15 @@ int32_t stm32_spi_init(struct no_os_spi_desc **desc,
 		if (ret)
 			goto error_pwm;
 	}
+	if (sinit->tx_pwm_init) {
```

Comment:
Hi @CiprianRegus 

The general sampling procedure for an ADC is:
- toggle CNV pin
- CS goes low
- Read spi words.

When trying to read multiple samples in standard SPI
mode there is significant latency setting up each write,
then polling or waiting treating the IRQ to and xfer data
from the SPI data register. This, for each 8bit spi word.

This latency could be reduced if DMA is used to read
directly words from the spi data register without cpu intervention.

However, we cannot simply program the TX DMA and RX DMA for N number of
transfers because:

- CNV pin needs to be toggled and synchronized with SPI read.
- CS is not automatically set between words (TI mode disabled)
- even if CS is set to auto: SPI does 8bit words or 16 bit words (max)
but some chips need more bits to be read between CS toggles.

Hence some other method of synchronizing CNV, CS, and SPI data read
needs to be implemented.

The method here uses STM32 timers to achieve a coordinated DMA transfer:
with stm32 timmers we can:

- configure the timer in pwm mode.
- master>slave hierarchy.
- one shot mode with repeat capability
- dma operations to peripheral. 

The proposed solution uses 3 timers and spi rx dma:

CNV timer:
 - job is to trigger the CNV pin and start CS timer
 - master to CS timer (event sent on timmer start)
 - period is sampling rate
 - duty cycle depends on part, typically 10ns.

 
CS timer
 - job is to wait for conversion ready, toggle CS and start TX timer
 - slave to CNV timer
 - master to TX timer (event sent on duty cycle complete)
 - one shot mode (one spi read each CNV toggle)
 - one repetition (CS stays low during n-byte read)
 - period is sampling rate
 - duty cycle is max busy time between CNV and data ready.
 
 TX timer
  - job is to write to SPI data register to write single 8 bit word to SPI
  - slave to CS timer
  - DMA enabled mem to peripheral mode (write to SPI->DR)
  - one shot mode
  - repetition is the number of bytes to transfer.
  - period is speed spi xfers 8bits IOW depends on spi baudrate
  - duty cycle dont care. we just use the start event.
 
 SPI RX DMA
  - triggered by SPI hw after each TX timer wrote to SPI->DR and SPI xfer is done.
  - SPI DMA xfers data to memory from SPI->DR

Reply:
Thank you for explaining this, I understand it now. Also, the timer synchronization mechanism sounds good in my opinion.